src/goto-instrument/contracts/contracts.cpp

```diff
@@ -140,123 +140,140 @@ void code_contractst::check_apply_loop_contracts(
+  //
+  // Asterisks (*) denote anchor points (goto targets) for instrumentations.
+  // We insert generated code above and/below these targets.
+  // Assertions for assigns clause checking are inserted in the loop body.
 
   // an intermediate goto_program to store generated instructions
-  goto_programt generated_code;
-
-  // process quantified variables correctly (introduce a fresh temporary)
-  // and return a copy of the invariant
-  const auto &invariant_expr = [&]() {
-    auto invariant_copy = invariant;
-    replace_symbolt replace;
-    code_contractst::add_quantified_variable(invariant_copy, replace, mode);
-    replace(invariant_copy);
-    return invariant_copy;
-  };
+  // to be inserted before the loop head
+  goto_programt pre_loop_head_instrs;
 
-  // Process "loop_entry" history variables
-  // Find and replace "loop_entry" expression in the "invariant" expression.
-  std::map<exprt, exprt> parameter2history;
+  // Process "loop_entry" history variables.
+  // We find and replace all "__CPROVER_loop_entry" subexpressions in invariant.
+  std::map<exprt, exprt> history_var_map;
   replace_history_parameter(
     invariant,
-    parameter2history,
-    loop_head->source_location(),
+    history_var_map,
+    loop_head_location,
     mode,
-    generated_code,
+    pre_loop_head_instrs,
     ID_loop_entry);
 
-  // Generate: assert(invariant) just before the loop
-  // We use a block scope to create a temporary assertion,
-  // and immediately convert it to goto instructions.
+  // Create a temporary to track if we enter a loop, i.e., get pas its guard.
+  const auto entered_loop =
+    new_tmp_symbol(
+      bool_typet(), loop_head_location, mode, symbol_table, "__entered_loop")
+      .symbol_expr();
+  pre_loop_head_instrs.add(
+    goto_programt::make_decl(entered_loop, loop_head_location));
+  pre_loop_head_instrs.add(
+    goto_programt::make_assignment(entered_loop, false_exprt{}));
+
+  // Create a snapshot of the invariant so that we can check the base case,
+  // if the loop is not vacuous and must be abstracted with contracts.
+  const auto initial_invariant_val =
+    new_tmp_symbol(
+      bool_typet(), loop_head_location, mode, symbol_table, "__init_invariant")
+      .symbol_expr();
+  pre_loop_head_instrs.add(
+    goto_programt::make_decl(initial_invariant_val, loop_head_location));
   {
-    code_assertt assertion{invariant_expr()};
-    assertion.add_source_location() = loop_head->source_location();
-    converter.goto_convert(assertion, generated_code, mode);
-    generated_code.instructions.back().source_location_nonconst().set_comment(
-      "Check loop invariant before entry");
+    // Although the invariant at this point will not have side effects,
+    // it is still a C expression, and needs to be "goto_convert"ed.
+    // Note that this conversion may emit many GOTO instructions,
+    // and we cannot simply do:
+    //    pre_loop_head_instrs.add(goto_programt::make_assignment(...))
```

Comment:
why?

Reply:
We should just remove the part "and we cannot simply do". The reason is already given above (the invariant expression is still a C expression and needs to be converted to a basic GOTO expression)